File: projeqtor-mcp-server/src/tools/tickets.ts
import { TicketPayloadSchema, IdSchema, CriteriaSchema, JsonObjectSchema } from "../schemas/projeqtor-types.js";
import { safeTool } from "../utils/error-handler.js";
import { mergeExtra, type ToolContext } from "./common.js";

export function registerTicketTools(server: any, ctx: ToolContext): void {
  server.registerTool("list_tickets", {
    title: "List tickets",
    description: "List ProjeQtOr tickets with optional filters: project, status, priority and responsible resource.",
    inputSchema: { projectId: IdSchema.optional(), statusId: IdSchema.optional(), priorityId: IdSchema.optional(), resourceId: IdSchema.optional() }
  }, (a: any) => safeTool(() => {
    const criteria = [
      ...(a.projectId ? [{ field: "idProject", operator: "=", value: a.projectId }] : []),
      ...(a.statusId ? [{ field: "idStatus", operator: "=", value: a.statusId }] : []),
      ...(a.priorityId ? [{ field: "idPriority", operator: "=", value: a.priorityId }] : []),
      ...(a.resourceId ? [{ field: "idResource", operator: "=", value: a.resourceId }] : [])
    ];
    return criteria.length ? ctx.client.search("Ticket", criteria) : ctx.client.listAll("Ticket");
  }));

  server.registerTool("create_ticket", {
    title: "Create ticket",
    description: "Create a ProjeQtOr Ticket for bug tracking/support. Write payload is AES-CTR encrypted.",
    inputSchema: TicketPayloadSchema.shape
  }, (a: any) => safeTool(() => ctx.client.create("Ticket", mergeExtra(a))));

  server.registerTool("update_ticket", {
    title: "Update ticket",
    description: "Update ticket status, priority, assignment or any ProjeQtOr ticket field.",
    inputSchema: { id: IdSchema, updates: JsonObjectSchema }
  }, (a: any) => safeTool(() => ctx.client.update("Ticket", { id: a.id, ...a.updates })));

  server.registerTool("search_tickets", {
    title: "Advanced ticket search",
    description: "Search Ticket objects with SQL-like criteria supported by ProjeQtOr's /search endpoint. Use for name contains, date ranges or custom fields.",
    inputSchema: { criteria: CriteriaSchema.array().min(1) }
  }, (a: any) => safeTool(() => ctx.client.search("Ticket", a.criteria)));
}

File: src/projeqtor_mcp_server/tools/tickets.py
from __future__ import annotations

from typing import Any, Annotated

from pydantic import Field
from mcp.server.fastmcp import FastMCP

from projeqtor_mcp_server.client.projeqtor_api import ProjeQtOrApiClient, SearchCriterion
from projeqtor_mcp_server.tools.common import Id, IdField, criteria_from_dicts, merge_extra, safe


def register_ticket_tools(mcp: FastMCP, client: ProjeQtOrApiClient) -> None:
    """Register ticket tools."""

    @mcp.tool(description="Lister les tickets avec filtres optionnels: projet, statut, priorité, responsable.")
    async def list_tickets(project_id: Id | None = None, status_id: Id | None = None, priority_id: Id | None = None, resource_id: Id | None = None) -> Any:
        criteria = []
        if project_id is not None:
            criteria.append(SearchCriterion("idProject", project_id))
        if status_id is not None:
            criteria.append(SearchCriterion("idStatus", status_id))
        if priority_id is not None:
            criteria.append(SearchCriterion("idPriority", priority_id))
        if resource_id is not None:
            criteria.append(SearchCriterion("idResource", resource_id))
        return await safe(client.search("Ticket", criteria) if criteria else client.list_all("Ticket"))

    @mcp.tool(description="Créer un ticket ProjeQtOr bug/support. Payload chiffré AES-CTR.")
    async def create_ticket(name: str, id_project: Id | None = None, id_ticket_type: Id | None = None, id_status: Id | None = None, id_priority: Id | None = None, id_resource: Id | None = None, description: str | None = None, extra: dict[str, Any] | None = None) -> Any:
        payload = merge_extra({"name": name, "idProject": id_project, "idTicketType": id_ticket_type, "idStatus": id_status, "idPriority": id_priority, "idResource": id_resource, "description": description}, extra)
        return await safe(client.create("Ticket", payload))

    @mcp.tool(description="Mettre à jour un ticket: statut, priorité, affectation ou champs natifs.")
    async def update_ticket(id: Annotated[Id, IdField], updates: dict[str, Any]) -> Any:
        return await safe(client.update("Ticket", {"id": id, **updates}))

    @mcp.tool(description="Recherche avancée de tickets via critères SQL-like: [{field, operator, value}].")
    async def search_tickets(criteria: Annotated[list[dict[str, Any]], Field(description="Critères: field, operator, value")]) -> Any:
        return await safe(client.search("Ticket", criteria_from_dicts(criteria)))
